Implement ShooterHoodSubsystem

FireShredder24 opened the pull request:
Provides a Subsystem for controlling the NEO-550 Brushless Motor that actuates the rack gear which controls the angle of deflection of Power Cells as they exit the shooter.
This Subsystem provides two types of control:
 - Fine, SmartMotion-enabled control intended for use in conjunction with vision
 - Binary, deploy-or-stow control intended for teleop.

FireShredder24 committed:
Create ShooterHoodSubsystem.java

FireShredder24 committed:
Add moveHood() with a variable moveRequest

FireShredder24 committed:
Setup ShooterHood for SmartMotion

FireShredder24 committed:
Setup ShooterHood for Position control

Fills out the SmartMotion configs, raiseHood() and lowerHood().

FireShredder24 committed:
Rename raise/lower methods to reflect direction

Renames raiseHood() and lowerHood() to reflect a higher sense of direction.
Adds a deploy state bool to the Subsystem.

FireShredder24 committed:
Expose deploy state

terbos90803 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -29,6 +29,7 @@
   public static final int SPARKMAX_SHOOTER_CAN_ADDR_RIGHT = 5;
   public static final int SPARKMAX_SHOOTER_CAN_ADDR_LEFT = 6;
   public static final int SPARKMAX_LOADER_CAN_ADDR = 7;
+  public static final int SPARKMAX_SHOOTER_HOOD_CAN_ADDR = 8;
```
The CAN addresses have not been selected for any of the SparkMAXs.  Sam is supposed to do this and up date the system chart at tonight's meeting.  Bug him until he gets this done.

terbos90803 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -29,6 +29,7 @@
+  public static final int SPARKMAX_SHOOTER_HOOD_CAN_ADDR = 8;
   // SparkMAX Encoder Channels
   public static final int SHOOTER_ENCODER_CHAN_A = 1;
   public static final int SHOOTER_ENCODER_CHAN_B = 2;
```
I'm not sure what encoder channels are.  I don't think it's what you want.  We should sort this out.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -0,0 +1,107 @@
+  private CANEncoder hoodEncoder = hoodNEO.getEncoder();
+  private CANPIDController hoodPID;
+
+  private final double kP = 5e-5;
+  private final double kI = 1e-6;
+  private final double kD = 0;
+  private final double kIz = 0;
+  private final double kFF = 0.000156;
+  private final double kMaxOutput = 0.3;
+  private final double kMinOutput = -0.3;
+  private final double allowedErr = 0;
```
These are the constants for the full NEO.  I've not seen any published constants for the NEO550.  So we can start with these, but we'll likely have to tune them.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -0,0 +1,107 @@
+
+  private double minVel = 0;
+  private double maxVel = 0;
+  private double maxAcc = 0;
```
These will need to be non-zero.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -0,0 +1,107 @@
+    hoodPID.setSmartMotionAllowedClosedLoopError(allowedErr, 0);
+
+    hoodNEO.getForwardLimitSwitch(CANDigitalInput.LimitSwitchPolarity.kNormallyOpen).enableLimitSwitch(false);
+    hoodLimitSwitch.enableLimitSwitch(enableLimit);
```
Redundant with line 49.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -0,0 +1,107 @@
+    hoodNEO.getForwardLimitSwitch(CANDigitalInput.LimitSwitchPolarity.kNormallyOpen).enableLimitSwitch(false);
+    hoodLimitSwitch.enableLimitSwitch(enableLimit);
+
+    hoodNEO.setIdleMode(IdleMode.kCoast);
```
Should be kBrake.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -0,0 +1,107 @@
+
+  private double hoodPos;
+  private boolean reliableZero;
+  private boolean enableLimit;
```
I'm not sure if we'd ever want to disable the limit after initial calibration.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -0,0 +1,107 @@
+  public void moveHood(double posRequest) {
+    // Used for autonomous and vision-tied control of the shooter hood.
+    hoodPID.setReference(posRequest, ControlType.kSmartMotion, 0);
+    enableLimit = false;
```
don't disable limit.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -0,0 +1,107 @@
+
+  public void deployHood() {
+    while (hoodPos < maxHoodPos) {
+      hoodPID.setReference(-0.1, ControlType.kSmartMotion, 0);
```
Never put while loops in code like this.  Blocking code like this will infinite loop and kill the robot.

The correct procedure is to simply set the state and return.

Keep in mind that setting the Reference is the same as setting any motor control.  You have to do it every cycle.  So again, this is the type of thing to save state here and then apply the state in the periodic method.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -0,0 +1,107 @@
+  }
+
+  public void stowHood() {
+    while (!hoodLimitSwitch.get()) {
```
Same as above.

FireShredder24 committed:
Merge branch 'master' into ShooterHood

FireShredder24 committed:
Refactor SHSS.Periodic()

Moves the setReference() call to ShooterHoodSubsystem.periodic() and sets its value based on a MoPref.

FireShredder24 committed:
Merge branch 'master' into ShooterHood

terbos90803 commented on src/main/java/frc/robot/Constants.java:
```diff
@@ -29,10 +29,9 @@
   public static final int SPARKMAX_SHOOTER_CAN_ADDR_RIGHT = 5;
   public static final int SPARKMAX_SHOOTER_CAN_ADDR_LEFT = 6;
   public static final int SPARKMAX_LOADER_CAN_ADDR = 7;
```
What is "Loader"

FireShredder24 replied:
I think I put that there while working on the shooter.  It's a deprecated version of the shooter gate, which is under StorageSubsystem

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -55,29 +55,23 @@ public ShooterHoodSubsystem() {
-    }
-    isDeployed = true;
+    deploy = true;
+    stow = false;
```
If deploy and stow are always opposites (which they are), they should never be represented by two different variables.

Options are:

- Single boolean called either stow OR deploy
- Enum indicating different state options, of which there are currently 2.

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -94,11 +88,17 @@ public void periodic() {
+    if (deploy)
+      hoodSetpoint = MoPrefs.getShooterHoodSetpoint();
+    if (stow)
+      hoodSetpoint = 0;
```
Cleanup for single controlling boolean.

FireShredder24 committed:
Change to single controlling boolean

terbos90803 commented on src/main/java/frc/robot/subsystems/ShooterHoodSubsystem.java:
```diff
@@ -91,7 +88,7 @@ public void periodic() {
     if (deploy)
       hoodSetpoint = MoPrefs.getShooterHoodSetpoint();
-    if (stow)
+    else
```
You've already initialized hoodSetpoint to 0 on 88.  No need to set it again in the else.

FireShredder24 committed:
Remove unnecessary re-init of hoodSetpoint
